JSON Exporter, part deux

gdpelican opened the pull request:
Thought I'd share this for a little feedback if you'd care to give it a look @jhass @jaywink

This currently:
- Boots the current export off to Sidekiq
- Uploads a json file through carrierwave
- Sends an email notifying the user when the job's complete, which includes a link to the upload
- Should pass tests, although we'll see when Travis gets its mitts on it.

It doesn't:
- Have a good UI solution yet (you just click on 'download my profile' and it takes you to a blank screen with status 200; we'll have to do a flash notice or something to tell the user what's happening)
- Include any of the heavy lifting associations we outscoped from the last PR

gdpelican committed:
Add contacts/posts, and GZipping JSON exporter output

gdpelican commented:
Alright, I've changed it so the data is GZipped and added in posts and comments to the serializers (might need some feedback on which fields want to be there)

It'd be cool to get your take on this, @jaywink ?

jaywink commented on app/models/user.rb:
```diff
@@ -291,6 +291,28 @@ def like_for(target)
+  end
+
+  def perform_export!
+    export = Tempfile.new([username, '.json'], encoding: 'ascii-8bit')
```
Hmm why is the encoding `ascii-8bit`?

jhass replied:
ASCII-8bit is Rubys clunky way of saying BINARY (which is an alias for ASCII-8bit in fact). Remember, ASCII is a 7bit encoding ;) Since we write gzipped data here, that's correct.

jaywink commented on app/controllers/users_controller.rb:
```diff
@@ -135,12 +135,14 @@ def getting_started_completed
+  end
+
+  def download_profile
+    send_data decompressed_export, type: :json, filename: current_user.export.filename
```
Is there any reason we can't just send over a compressed file? IMHO as it's text data, this would make sense to decrease downloads. I'm guessing it will be 1 week after this is released that someone will offer an utility to automate exporting of profiles.. ;)

gdpelican replied:
Just send it over as `application/gzip` with a `.gz` file extension, and let the user handle inflating it?

I worry about people on windows machines needing to download a utility to open that file, although I guess we could look at providing a .zip format as well.

jhass replied:
Or maybe make it a tarball, even most Windows users should have a tool for that, I think 7-zip is quite common and can handle those.

jaywink commented on app/serializers/export/comment_serializer.rb:
```diff
@@ -0,0 +1,8 @@
+module Export
+  class CommentSerializer < ActiveModel::Serializer
+    attributes :text,
+               :author_id,
```
`author_id` and `author_signature` are probably irrelevant for a user, even in a pod migration situation (I think). Could we instead add here author diaspora handle? That is the network wide unique identifier that is actually usable. GUID is of course too but that doesn't say anything to users looking at the export.

jaywink replied:
Heh no wait, this is of course the users own comments, so no need to have a handle. BUT we need to say what post it relates to. How about building an url to the post?

jhass replied:
I'd just export the post guid.

jaywink replied:
Post guid is good, agreed. It can be used to make a link anyway :)

jaywink commented on app/serializers/export/post_serializer.rb:
```diff
@@ -0,0 +1,16 @@
+module Export
+  class PostSerializer < ActiveModel::Serializer
+    attributes :text,
+               :author_id,
```
Again, `author_id` is probably irrelevant.

jaywink commented on app/serializers/export/post_serializer.rb:
```diff
@@ -0,0 +1,16 @@
+               :likes_count,
+               :comments_count,
+               :reshares_count,
+               :favorite
```
Interesting, first time seen this and doesn't to me look like it's used. There is a `toggleFavorite` method in the JS but it's not called :P

Some legacy idea that never got implemented, @jhass ? I'd personally remove it if we don't use the attribute.

jhass replied:
Eh, I don't remember that one at all, must be from the time I didn't watch the project. It was introduced in 75d25e947502851e2b79a6f9a101a2a83d2eb363

jaywink commented on app/serializers/export/comment_serializer.rb:
```diff
@@ -0,0 +1,8 @@
+    attributes :text,
+               :author_id,
+               :author_signature,
+               :likes_count
```
We don't support `likes_count` at the moment so this could be removed.

jhass replied:
More importantly it's just a counter cache.

jaywink commented:
Other than the comments relating to the attributes, **super awesome** work :star2: A lot of people will be happy to have this working again, as it's a shame our export is currently worse than many proprietary networks ;)

:sake: for you ;)

jaywink commented:
Oh also, I'm assuming this will work fine with pods using S3 for storage etc?

jhass commented on app/serializers/export/post_serializer.rb:
```diff
@@ -0,0 +1,16 @@
+               :image_width,
+               :likes_count,
+               :comments_count,
+               :reshares_count,
```
Counter caches, shouldn't be included.

jaywink replied:
Doesn't it contain the current (when exported) count though? IMHO that is relevant information for the user :)

jhass replied:
It should, I don't see much value in that information though.

jaywink replied:
I think it could be valuable for some - imagine you want to look at some old export of your data from an old pod for example. Maybe someone will make an app to analyse exports and rank posts with most likes etc.

jhass commented on app/uploaders/exported_user.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  def filename
+    :"#{model.username}_diaspora_data.json"
```
Why does this need to be a Symbol?

gdpelican replied:
It doesn't, I'll change it.

jaywink commented:
Also, before merging, the changelog entry for the export should probably be refreshed. I guess the export version does not need to be bumped since the current export version hasn't gone stable yet.

gdpelican commented:
I haven't been able to test using an S3 instance, but I just aped the Carrierwave uploaders for images, so it shouldn't behave any differently.

gdpelican commented:
Well, I b0rked the rebase a little, but merged in `develop`, updated the serializers, and changed that one symbol that wasn't supposed to be there.

Also made it so we export files in a .json.gz file, which 7zip should handle.

jaywink commented:
Alright, no more comments to this and looks awesome to me - merging! Nice to get this in to the next release, thanks @gdpelican !

Flaburgan commented:
Awesome!

Flaburgan commented:
Currently in test on diaspora-fr :)
![sidekiq-export](https://cloud.githubusercontent.com/assets/930064/5786194/6d18158c-9de1-11e4-8b7d-64a884ccdf18.png)

marienfressinaud commented:
@Flaburgan > I tried yesterday night but it still shows me "We are currently processing your data. Please check back in a few moments." this morning. I'm pretty sure there is a problem ;).

gdpelican commented:
Oh noes! I'll have a poke at it tomorrow to see what's up. We're sure
sidekiq is running?

On Sat, Jan 17, 2015 at 11:22 PM, Marien Fressinaud <
notifications@github.com> wrote:

> @Flaburgan <https://github.com/Flaburgan> > I tried yesterday night but
> it still shows me "We are currently processing your data. Please check back
> in a few moments." this morning. I'm pretty sure there is a problem ;).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diaspora/diaspora/pull/5499#issuecomment-70361897>.
>

Flaburgan commented:
Yeah sidekiq is running, Marien and my jobs are listed in "Enqueued". They are the only two jobs in the queue.

jhass commented:
`export_user` wasn't added to https://github.com/diaspora/diaspora/blob/develop/config/sidekiq.yml#L6

Flaburgan commented:
Confirmed, adding export_user there immediately made the jobs processed. But the message on the panel is still `We are currently processing your data. Please check back in a few moments.`

marienfressinaud commented:
Indeed but I have a different message: ` download my profile (Last updated at 2015-01-17 12:36:04 UTC)
refresh my profile data ` so it seems ok for me. Nevertheless there are other bugs, should we open new specific tickets for each bug found?

jhass commented:
> should we open new specific tickets for each bug found?

Yes, sure.

jaywink commented:
@gdpelican the sidekiq queue thingy is fixed now, no need to check that :)

margori commented:
I have a doubt in how sidekiq works.
- I configured sidekiq embed in unicorn in my local dev pod. (embed_sidekiq_worker: true in diaspora.yml)
-when I click over "refresh my profile" link, I expect to have new registers in 'sidekiq.log' file.

Am I wrong?
